### PR TITLE
[WIP] feat(iroh-net): PCP/NAT-PMP

### PR DESCRIPTION
**Pull request by divagant-martian**
## Description

Implement probing and mapping for PCP/NAT-PMP

## Notes & open questions

TBD

## Change checklist

- [ ] Self-review.
- [ ] Documentation updates if relevant.
- [ ] Tests if relevant.

### CODE REVIEW

File: iroh-net/src/hp/portmapper/pmp.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+const MIN_RESP_SIZE: usize = 1 + 1 + 2 + 4 + 4;
+/// Minimum size of an encoded [`Response`] sent by a server to this client.
+// NOTE: 1byte for the version +
+//       1byte for the opcode +
+//       2byte for the result code +
+//       4bytes for the epoch time +
+//       2bytes for the private port +
+//       2bytes for the public port +
+//       4bytes for the lifetime = response size for a mapping request
+const MAX_RESP_SIZE: usize = 1 + 1 + 2 + 4 + 2 + 2 + 4;
```
Comment (dignifiedquire) [inline iroh-net/src/hp/portmapper/pmp.rs:16]: Maybe a bit more readable

```rust
const MAX_RESP_SIZE: usize = 1  // Version
+ 1 // Opcode
+ 2 // Result Code
+ 4 // Epoch time
+ 2 // Private Port
+ 2 // Public Port
+ 4; // the lifetime = response size for a mapping request
```

File: iroh-net/src/hp/portmapper/pmp.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+                lifetime_seconds,
+            } => {
+                let opcode = match proto {
+                    MapProtocol::UDP => Opcode::MapUdp,
+                    MapProtocol::TCP => Opcode::MapTcp,
+                };
+                let mut buf = vec![Version::NatPmp as u8, opcode as u8];
+                buf.extend_from_slice(&local_port.to_be_bytes());
+                buf.extend_from_slice(&external_port.to_be_bytes());
+                buf.extend_from_slice(&lifetime_seconds.to_be_bytes());
```
Comment (dignifiedquire) [inline iroh-net/src/hp/portmapper/pmp.rs:97]: I wish rust had a syntax like this..

```rust
let buf = [
  Version::NatPmp as u8, 
  opcode as u8,
  local_port.to_be_bytes()..,
  external_port.to_be_bytes()..,
  lifetime_seconds.to_be_bytes()..,
];
```

File: iroh-net/src/hp/portmapper/pmp.rs [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+/// Error ocurring when attempting to decode the [`ResultCode`] in a server response.
+#[derive(Debug, PartialEq, Eq)]
+pub struct InvalidResultCode;
+
+impl TryFrom<u16> for ResultCode {
+    type Error = InvalidResultCode;
+
+    fn try_from(value: u16) -> std::result::Result<Self, Self::Error> {
```
Comment (dignifiedquire) [inline iroh-net/src/hp/portmapper/pmp.rs:158]: This should make some of this code a little easier to write :) 
https://docs.rs/num_enum/

**Comment by divagant-martian** [comment]
I'm going to split this PR in four because it's way too big. After splitting the first one and doing the work of splitting the second one found a bug, so definitely this split helps with doing proper reviews

**Pull request closed without merging**